controller/orphan_controller.go

```diff
@@ -0,0 +1,387 @@
+			orphan.Name, oc.controllerID)
+		return nil
+	}
+
+	if types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeDataCleanable).Status ==
+		longhorn.ConditionStatusFalse {
+		log.Infof("Only delete orphan %v resource object and do not delete the orphaned data", orphan.Name)
+		return nil
+	}
+
+	switch orphan.Spec.Type {
+	case longhorn.OrphanTypeReplica:
+		err = oc.deleteOrphanedReplica(orphan)
+	}
+
+	if err == nil || (err != nil && apierrors.IsNotFound(err)) {
+		return nil
+	}
+
+	return err
+}
+
+func (oc *OrphanController) deleteOrphanedReplica(orphan *longhorn.Orphan) error {
+	oc.logger.Infof("Deleting orphan %v replica directory %v in disk %v on node %v",
+		orphan.Name, orphan.Spec.Parameters[longhorn.OrphanDataName],
+		orphan.Spec.Parameters[longhorn.OrphanDiskPath], orphan.Status.OwnerID)
+
+	return util.DeleteReplicaDirectoryName(orphan.Spec.Parameters[longhorn.OrphanDiskPath], orphan.Spec.Parameters[longhorn.OrphanDataName])
+}
+
+func (oc *OrphanController) updateConditions(orphan *longhorn.Orphan) error {
+	if err := oc.updateUnavailableCondition(orphan); err != nil {
+		return err
+	}
+
+	if types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeError).Status != longhorn.ConditionStatusTrue {
+		orphan.Status.Conditions = types.SetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeError, longhorn.ConditionStatusFalse, "", "")
+	}
+
+	return nil
+}
+
+// If the Unavailable condition status is ture, then the deletion of the orphan CRD won't delete the on-disk data
+func (oc *OrphanController) updateUnavailableCondition(orphan *longhorn.Orphan) (err error) {
+	reason := ""
+
+	defer func() {
+		if err != nil {
+			return
+		}
+
+		status := longhorn.ConditionStatusTrue
+		if reason != "" {
+			status = longhorn.ConditionStatusFalse
+		}
+
+		orphan.Status.Conditions = types.SetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeDataCleanable, status, reason, "")
+	}()
+
+	node, err := oc.ds.GetNodeRO(orphan.Spec.NodeID)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get node %v", orphan.Spec.NodeID)
+		}
+		reason = longhorn.OrphanConditionTypeDataCleanableReasonNodeUnavailable
+		return nil
+	}
+
+	isUnavailable, err := oc.ds.IsNodeDownOrDeletedOrMissingManager(orphan.Spec.NodeID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get node %v", orphan.Spec.NodeID)
+	}
+	if isUnavailable {
+		reason = longhorn.OrphanConditionTypeDataCleanableReasonNodeUnavailable
+		return nil
+	}
+
+	if node.Spec.EvictionRequested {
+		reason = longhorn.OrphanConditionTypeDataCleanableReasonNodeEvicted
+		return nil
+	}
+
+	switch {
+	case orphan.Spec.Type == longhorn.OrphanTypeReplica:
+		reason = checkOrphanedReplicaUnavailable(node, orphan)
+	}
+
+	return nil
+}
+
+func checkOrphanedReplicaUnavailable(node *longhorn.Node, orphan *longhorn.Orphan) string {
+	id := orphan.Spec.Parameters[longhorn.OrphanDiskFsid]
```

Comment:
Should use `DiskUUID` instead. 
BTW, it's better to name the disk-related fields as `id`. Instead, using `fsid`, `diskName`, or `diskUUID`.

Reply:
weird, this is old code...Why isn't the code updated ? :(